### ISSUE TITLE
[BUG] Setuptools causes failure when parsing environment marker with parentheses

### ISSUE THREAD

**Issue by amarkules1** [pypa/setuptools#3840]
### setuptools version

setuptools >= 67.0.0

### Python version

Python 3.9, Python 3.10, likely others

### OS

Windows 10, Ubuntu 20.04, likely others

### Description

Ran into an error after upgrading to setuptools 67.4.0 and trying to build the wheel files for a project. It appears the new version is causing some issue when parsing an environment marker. The requirements.txt file I believe it is reading from has a space and right parenthesis after the "or" where the log seems to indicate a missing character. Editing the Pipfile to use an earlier version of setuptools fixes the issue.

### Expected behavior

I expect setuptools to interpret the environment marker correctly, and to not prevent me from building wheel files for my projects

### How to Reproduce

1. Clone https://github.com/amarkules1/setuptools-bug-example.git
2. Run build_wheels.sh (assumes python3 and pipenv are installed)
3. Observe output in console.

### Output

```console
./build_wheels.sh
Pipfile.lock not found, creating...
Locking [packages] dependencies...
           Building requirements...
Resolving dependencies...
Success!
Locking [dev-packages] dependencies...
Updated Pipfile.lock (c80165ac5802f8af0a9f6de3928ddce43d36ad09f1fa4f37384dd7d5826d8505)!
Installing dependencies from Pipfile.lock (6d8505)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
Processing c:\dev\setuptools-bug-example
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [54 lines of output]
      [pbr] Generating ChangeLog
      C:\dev\setuptools-bug-example\.venv\lib\site-packages\setuptools\command\easy_install.py:144: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      C:\dev\setuptools-bug-example\.venv\lib\site-packages\setuptools\command\install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      Marker evaluation failed, see the following error.  For more information see: http://docs.openstack.org/pbr/latest/user/using.html#environment-markers
      Error parsing
      Traceback (most recent call last):
        File "C:\dev\setuptools-bug-example\.venv\lib\site-packages\pkg_resources\_vendor\packaging\markers.py", line 192, in __init__
          self._markers = _normalize_extra_values(parse_marker(marker))
        File "C:\dev\setuptools-bug-example\.venv\lib\site-packages\pkg_resources\_vendor\packaging\_parser.py", line 234, in parse_marker
          return _parse_marker(Tokenizer(source, rules=DEFAULT_RULES))
        File "C:\dev\setuptools-bug-example\.venv\lib\site-packages\pkg_resources\_vendor\packaging\_parser.py", line 241, in _parse_marker
          expression = [_parse_marker_atom(tokenizer)]
        File "C:\dev\setuptools-bug-example\.venv\lib\site-packages\pkg_resources\_vendor\packaging\_parser.py", line 260, in _parse_marker_atom
          tokenizer.consume("WS")
        File "C:\Users\amarkules\AppData\Local\Programs\Python\Python39\lib\contextlib.py", line 126, in __exit__
          next(self.gen)
        File "C:\dev\setuptools-bug-example\.venv\lib\site-packages\pkg_resources\_vendor\packaging\_tokenizer.py", line 183, in enclosing_tokens
          self.raise_syntax_error(
        File "C:\dev\setuptools-bug-example\.venv\lib\site-packages\pkg_resources\_vendor\packaging\_tokenizer.py", line 163, in raise_syntax_error
          raise ParserSyntaxError(
      pkg_resources.extern.packaging._tokenizer.ParserSyntaxError: Expected closing RIGHT_PARENTHESIS
          (platform_machine == 'aarch64' orplatform_machine == 'ppc64le' or (platform_machine == 'x86_64' or (platform_machine == 'amd64' or (platform_machine == 'AMD64' or (platform_machine == 'win32' or platform_machine == 'WIN32')))
))
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
     
      The above exception was the direct cause of the following exception:
     
      Traceback (most recent call last):
        File "C:\dev\setuptools-bug-example\.venv\lib\site-packages\pkg_resources\__init__.py", line 1432, in evaluate_marker
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

```

**Comment by abravalheri** [comment]
To continue investigating this, I think we need a more simplified repro that does not rely in `pbr` and is able to reproduce the error for the latest version of setuptools.

**Comment by abravalheri** [comment]
I also noticed in your example that the error `pip` shows seems to be missing an `or` right after the first environment marker:

```console
      pkg_resources.extern.packaging._tokenizer.ParserSyntaxError: Expected closing RIGHT_PARENTHESIS
          (platform_machine == 'aarch64' orplatform_machine == 'ppc64le' or (platform_machine == 'x86_64' or (platform_machine == 'amd64' or (platform_machine == 'AMD64' or (platform_machine == 'win32' or platform_machine == 'WIN32')))
))
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
```

This does not seem to match the contents of the [`requirements.txt` file](https://github.com/amarkules1/setuptools-bug-example/blob/main/requirements.txt) linked in the first post.

Any chance the error report and the reproducer may be diverging?

**Comment by amarkules1** [comment]
Hi @abravalheri, interesting that the error doesn't seem to happen without pbr. I had assumed that setuptools was to blame because the error doesn't happen when I change the version of setuptools to 66.1.1, but keep the version of pbr the same. I'm not entirely sure how, but maybe it's some combination of the 2 that's causing the issue?

As for your last comment, I noticed the missing space and right parenthesis in the error trace as well. My best guess is that these characters are getting removed somehow, at some point after they're read from requirements.txt, and that in turn leads to the error I'm seeing.

I did just notice however that the [requires.txt file](https://github.com/amarkules1/setuptools-bug-example/blob/main/setuptools_bug_example.egg-info/requires.txt) appears to have the same 2 characters missing. This is the case both with setuptools version 67.0.0 and 66.1.1, even though no error is thrown in the older version. That seems to indicate that the issue is with pbr, and that the earlier version of setuptools was just ignoring the invalid environment marker. I'll have look deeper to figure out what exactly is going on, but since it seems likely that the issue is with pbr, not setuptools, I'll close out this bug report. Thank you for your time looking into this.
